src/main/java/com/aws/greengrass/localdebugconsole/DashboardServer.java

```diff
@@ -212,6 +225,36 @@ public void onMessage(WebSocket conn, String msg) {
                     sendIfOpen(conn, new Message(MessageType.RESPONSE, packedRequest.requestID, true));
                     break;
                 }
+                case subscribeToPubSubTopic: {
+                    if (!pubSubWatchList.get(conn).containsKey(req.args[0])) {
+                        Consumer<PublishEvent> cb = (c) -> {
+                            CommunicationMessage resMessage = new CommunicationMessage(
+                                    req.args[0],
+                                    c.getTopic(),
+                                    new String(c.getPayload()));
+                            sendIfOpen(conn, new Message(MessageType.PUB_SUB_MSG, resMessage));
+                        };
+                        SubscribeRequest subReq =
+                                SubscribeRequest.builder().callback(cb).receiveMode(ReceiveMode.RECEIVE_ALL_MESSAGES)
+                                        .topic(req.args[0]).serviceName(SERVICE_NAME).build();
+                        pubSubIPCAgent.subscribe(subReq);
+                        pubSubWatchList.get(conn).put(req.args[0], subReq);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Your code calls `ConcurrentHashMap.containsKey`. If `containsKey` returns `false`, your code then calls `ConcurrentHashMap.put`. Because these operations are not atomic, the key could already exist in the map when you call `put()`, even if `containsKey` returns `false`. This could lead to data loss. Instead of `put()`, call `putIfAbsent()`, which associates an unassociated key with a value atomically. See the following diffs on GitHub that make this code change: [diff1](https://github.com/ninjaframework/ninja/commit/921cbd8b05a4ea930a5a5ad7e451a53e15ec03b7#diff-f39d2fcbd1ca7bbafe398d2a9aaeaf5edd8a23661c40b2f82351ff79ac1b23aeL121), [diff2](https://github.com/kieker-monitoring/kieker/commit/43c9ebd98a4597ea71021c96d04e125ee589d7f8#diff-053788507d3e6fdb69e8ca94e5b2bd0a86b03e3cd2d53a452e18dda69ea071d0L152)

Learn more: [ConcurrentHashMap.putIfAbsent in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#putIfAbsent-K-V-)